cmd/common/st/transfer.go

```diff
@@ -0,0 +1,307 @@
+	}
+	return nil
+}
+
+func transferFiles(source, target storage.Folder, parentDirPath string) error {
+	files, err := listFilesToMove(source, target, parentDirPath)
+	if err != nil {
+		return err
+	}
+
+	fileStates := make(chan transferFileState, len(files))
+	for _, f := range files {
+		fileStates <- transferFileState{
+			path: f.GetName(),
+		}
+	}
+
+	workersNum := utility.Min(transferConcurrency, len(files))
+	errs := make(chan error, workersNum)
+
+	workersWG := new(sync.WaitGroup)
+	workersWG.Add(workersNum)
+	for i := 0; i < workersNum; i++ {
+		go moveFilesWorker(source, target, fileStates, errs, workersWG)
+	}
+
+	errsWG := new(sync.WaitGroup)
+	errsWG.Add(1)
+	errsNum := 0
+	go func() {
+		defer errsWG.Done()
+		for e := range errs {
+			errsNum++
+			tracelog.ErrorLogger.PrintError(e)
+		}
+	}()
+
+	workersWG.Wait()
+	close(errs)
+	errsWG.Wait()
+
+	if errsNum > 0 {
+		return fmt.Errorf("finished with %d errors", errsNum)
+	}
+
+	return nil
+}
+
+func listFilesToMove(source, target storage.Folder, parentDirPath string) ([]storage.Object, error) {
+	isSubDir := func(dirPath string) bool {
+		return strings.HasPrefix(dirPath, parentDirPath)
+	}
+
+	targetFiles, err := storage.ListFolderRecursivelyWithFilter(target, isSubDir)
+	if err != nil {
+		return nil, fmt.Errorf("can't list files in the target storage: %w", err)
+	}
+	sourceFiles, err := storage.ListFolderRecursivelyWithFilter(source, isSubDir)
+	if err != nil {
+		return nil, fmt.Errorf("can't list files in the source storage: %w", err)
+	}
+	tracelog.InfoLogger.Printf("Total files in the source storage: %d", len(sourceFiles))
+
+	missingFiles := make(map[string]storage.Object, len(sourceFiles))
+	for _, sourceFile := range sourceFiles {
+		missingFiles[sourceFile.GetName()] = sourceFile
+	}
+	for _, targetFile := range targetFiles {
+		if transferOverwrite {
+			sourceFile := missingFiles[targetFile.GetName()]
+			if sourceFile.GetSize() != targetFile.GetSize() {
+				tracelog.WarningLogger.Printf(
+					"File present in both storages and its size is different: %q (source %d bytes VS target %d bytes)",
+					targetFile.GetName(),
+					sourceFile.GetSize(),
+					targetFile.GetSize(),
+				)
+			}
+		} else {
+			delete(missingFiles, targetFile.GetName())
+		}
+	}
+	tracelog.InfoLogger.Printf("Files missing in the target storage: %d", len(missingFiles))
+
+	if transferMax < 0 {
+		transferMax = math.MaxInt
+	}
+	count := 0
+	limitedFiles := make([]storage.Object, 0, utility.Min(transferMax, len(missingFiles)))
+	for _, file := range missingFiles {
+		if count >= transferMax {
+			break
+		}
+		limitedFiles = append(limitedFiles, file)
+		count++
+	}
+	tracelog.InfoLogger.Printf("Files will be transferred: %d", len(limitedFiles))
+
+	return limitedFiles, nil
+}
+
+type transferFileState struct {
+	path            string
+	copied          bool
+	prevCheck       time.Time
+	performedChecks uint
+}
+
+func moveFilesWorker(
```

Comment:
At first, we create several workers, all of which are listening to the `fileStates` channel. This channel initially has all the files to move.

Each worker has an infinite loop inside and processes a single file on each iteration. 
However, file processing is split into several stages. The files pulled from the channel on each iteration may be in different stages - the worker decides what to do with this file depending on its `transferFileState`.

1. On the first processing, the file is only written to the target storage. After that, it's enqueued to the end of the channel again
2. On the second processing, the worker checks if the file has appeared in the target storage. If so, it's deleted from the source storage and the state isn't enqueued to the channel again. If it hasn't appeared yet, the worker puts the file again at the end of the channel queue.

We wait until files appeared in the target storage before deleting them, in order to eliminate the risk of causing a moment of time when the file isn't visible in both storages - source and target. This situation is possible if the storage doesn't guarantee read-after-write consistency.

We process files in several steps and pass them through the same channel several times, to easily guarantee that the concurrency limit isn't exceeded: the required number of workers is created once only at the very beginning.

Please note that worker finishes when there are no more files in the channel. This situation is possible even if not all the files have been already moved, because some files may be being processed by other workers at the moment. It's not a problem, because in any case at every moment we have at least as many workers as there are files in progress.